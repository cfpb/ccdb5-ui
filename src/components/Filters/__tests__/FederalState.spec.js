import React from 'react';
import configureMockStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import ReduxFederalState, {
  FederalState,
  mapDispatchToProps,
} from '../FederalState';

const fixture = [
  {
    key: 'DC',
    doc_count: 999,
  },
];

/**
 *
 */
function setupEnzyme() {
  const props = {
    options: fixture,
    forTypeahead: [
      { key: 'AZ', label: 'Arizona (AZ)', normalized: 'arizona (az)' },
      { key: 'CO', label: 'Colorado (CO)', normalized: 'colorado (co)' },
      { key: 'CT', label: 'Connecticut (CT)', normalized: 'connecticut (ct)' },
      { key: 'MD', label: 'Maryland (MD)', normalized: 'maryland (md)' },
      {
        key: 'RI',
        label: 'Rhode Island (RI)',
        normalized: 'rhode island (ri)',
      },
      { key: 'WY', label: 'Wyoming (WY)', normalized: 'wyoming (wy)' },
    ],
    selections: ['DC'],
    hasChildren: true,
    typeaheadSelect: jest.fn(),
  };

  const target = shallow(<FederalState {...props} />);

  return {
    props,
    target,
  };
}

/**
 *
 * @param initialFixture
 */
function setupSnapshot(initialFixture) {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const store = mockStore({
    query: {
      state: ['DC'],
    },
    aggs: {
      state: initialFixture,
    },
  });

  return renderer.create(
    <Provider store={store}>
      <IntlProvider locale="en">
        <ReduxFederalState />
      </IntlProvider>
    </Provider>
  );
}

describe('component::FederalState', () => {
  describe('initial state', () => {
    it('renders empty values without crashing', () => {
      const target = setupSnapshot();
      const tree = target.toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders without crashing', () => {
      const target = setupSnapshot(fixture);
      const tree = target.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Typeahead interface', () => {
    let target, props;
    beforeEach(() => {
      ({ target, props } = setupEnzyme());
    });

    describe('_onInputChange', () => {
      it('matches state abbreviations first', () => {
        const actual = target.instance()._onInputChange('RI');
        expect(actual.length).toEqual(2);
        expect(actual[0].key).toEqual('RI');
        expect(actual[1].key).toEqual('AZ');
      });

      it('matches state abbreviations first (for code coverage)', () => {
        const actual = target.instance()._onInputChange('CO');
        expect(actual.length).toEqual(2);
        expect(actual[0].key).toEqual('CO');
        expect(actual[1].key).toEqual('CT');
      });

      it('prefers matches early in the string', () => {
        const actual = target.instance()._onInputChange('AR');
        expect(actual.length).toEqual(2);
        expect(actual[0].key).toEqual('AZ');
        expect(actual[1].key).toEqual('MD');
      });
    });

    describe('_renderOption', () => {
      it('produces a custom component', () => {
        const options = target.instance()._onInputChange('ARI');
        const view = target.instance()._renderOption(options[0]);
        expect(view).toEqual({
          value: 'AZ',
          component: expect.anything(),
        });
      });
    });

    describe('_onOptionSelected', () => {
      it('checks all the filters associated with the option', () => {
        const key = 'WY';
        target.instance()._onOptionSelected({ key });
        expect(props.typeaheadSelect).toHaveBeenCalledWith(key);
      });
    });
  });

  describe('StickyOptions interface', () => {
    it('provides the correct label for a missing aggregation', () => {
      const { target } = setupEnzyme();
      const key = 'WY';
      const actual = target.instance()._onMissingItem(key);
      expect(actual).toEqual({
        key,
        doc_count: 0,
        value: 'Wyoming (WY)',
      });
    });
  });

  describe('mapDispatchToProps', () => {
    it('hooks into addMultipleFilters', () => {
      const dispatch = jest.fn();
      mapDispatchToProps(dispatch).typeaheadSelect('baz');
      expect(dispatch.mock.calls.length).toEqual(1);
    });
  });
});
