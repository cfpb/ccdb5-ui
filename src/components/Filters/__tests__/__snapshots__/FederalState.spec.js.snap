// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component::FederalState initial state renders empty values without crashing 1`] = `
<section
  className="o-expandable aggregation state"
>
  <div
    className="o-expandable_header o-expandable_target"
  >
    <h4
      className="o-expandable_header-left o-expandable_label"
    >
      State
    </h4>
    <span
      className="o-expandable_header-right o-expandable_link"
    >
      <button
        aria-label="Hide State filter"
        className="a-btn a-btn__link o-expandable_cue o-expandable_cue-close"
        onClick={[Function]}
      >
        Hide
        <svg
          className="cf-icon-svg"
          viewBox="0 0 1000 1200"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M500 105.2c-276.1 0-500 223.9-500 500s223.9 500 500 500 500-223.9 500-500-223.9-500-500-500zm263.1 550.7H236c-27.6 0-50-22.4-50-50s22.4-50 50-50h527.1c27.6 0 50 22.4 50 50s-22.4 50-50 50z"
          />
        </svg>
      </button>
    </span>
  </div>
  <p>
    The state in the mailing address provided by the consumer
  </p>
  <section
    className="typeahead "
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <div
      className="m-btn-inside-input input-contains-label"
    >
      <div
        className="input-contains-label_before input-contains-label_before__search"
      >
        <svg
          className="cf-icon-svg"
          viewBox="0 0 880.6 1200"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M860.1 879.2L645.7 664.8c90.8-136.8 76-323-44.6-443.5-137.5-137.5-360.5-137.5-498 0s-137.5 360.5 0 498c118.5 118.4 303.9 137 443.5 44.6L761 978.3c27.3 27.3 71.7 27.3 99 0s27.4-71.8.1-99.1zm-508-116.9C191.4 762 61.3 631.5 61.6 470.8c.3-160.7 130.8-290.8 291.5-290.5s290.8 130.8 290.5 291.5c-.2 118.2-71.9 224.6-181.5 269.1-34.9 14.2-72.3 21.5-110 21.4z"
          />
        </svg>
      </div>
      <label
        className="u-visually-hidden"
        htmlFor="state-typeahead"
      >
        Start typing to begin listing US states
      </label>
      <input
        autoComplete="off"
        className="a-text-input"
        id="state-typeahead"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Enter state name or abbreviation"
        type="text"
        value=""
      />
    </div>
  </section>
  <ul />
</section>
`;

exports[`component::FederalState initial state renders without crashing 1`] = `
<section
  className="o-expandable aggregation state"
>
  <div
    className="o-expandable_header o-expandable_target"
  >
    <h4
      className="o-expandable_header-left o-expandable_label"
    >
      State
    </h4>
    <span
      className="o-expandable_header-right o-expandable_link"
    >
      <button
        aria-label="Hide State filter"
        className="a-btn a-btn__link o-expandable_cue o-expandable_cue-close"
        onClick={[Function]}
      >
        Hide
        <svg
          className="cf-icon-svg"
          viewBox="0 0 1000 1200"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M500 105.2c-276.1 0-500 223.9-500 500s223.9 500 500 500 500-223.9 500-500-223.9-500-500-500zm263.1 550.7H236c-27.6 0-50-22.4-50-50s22.4-50 50-50h527.1c27.6 0 50 22.4 50 50s-22.4 50-50 50z"
          />
        </svg>
      </button>
    </span>
  </div>
  <p>
    The state in the mailing address provided by the consumer
  </p>
  <section
    className="typeahead "
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <div
      className="m-btn-inside-input input-contains-label"
    >
      <div
        className="input-contains-label_before input-contains-label_before__search"
      >
        <svg
          className="cf-icon-svg"
          viewBox="0 0 880.6 1200"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M860.1 879.2L645.7 664.8c90.8-136.8 76-323-44.6-443.5-137.5-137.5-360.5-137.5-498 0s-137.5 360.5 0 498c118.5 118.4 303.9 137 443.5 44.6L761 978.3c27.3 27.3 71.7 27.3 99 0s27.4-71.8.1-99.1zm-508-116.9C191.4 762 61.3 631.5 61.6 470.8c.3-160.7 130.8-290.8 291.5-290.5s290.8 130.8 290.5 291.5c-.2 118.2-71.9 224.6-181.5 269.1-34.9 14.2-72.3 21.5-110 21.4z"
          />
        </svg>
      </div>
      <label
        className="u-visually-hidden"
        htmlFor="state-typeahead"
      >
        Start typing to begin listing US states
      </label>
      <input
        autoComplete="off"
        className="a-text-input"
        id="state-typeahead"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Enter state name or abbreviation"
        type="text"
        value=""
      />
    </div>
  </section>
  <ul>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="DC"
        checked={true}
        className="flex-fixed a-checkbox"
        id="state-dc"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="state-dc"
      >
        District Of Columbia (DC)
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          999
        </span>
      </span>
    </li>
  </ul>
</section>
`;
